sql/bulkingest: add range split and scatter utilities

spilchen opened the pull request:
Adds split and scatter utilities for distributed merge from the prototype branch. These utilities prepare the KV keyspace for efficient bulk ingest by pre-splitting ranges at SST boundaries and distributing them across cluster nodes.

Closes #156574
Epic: CRDB-48845
Release note: none

Co-authored by: @jeffswenson

mw5h reviewed (changes requested):
> @mw5h reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson and @kev-cao)

___
*[`pkg/sql/bulkingest/split_picker.go` line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156877#-OdKyBxt27GbQCGRXbxr:-OdKyBxt27GbQCGRXbxs:b-b5w0to) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bb58a611be76c7dfc26a18379031ef673d00485/pkg/sql/bulkingest/split_picker.go#L108)):*
> ```Go
> 		result = append(result, roachpb.Span{
> 			Key:    spanStart,
> 			EndKey: roachpb.Key(ssts[i].StartKey),
> ```

nit: maybe assign this value to a variable "splitPoint" or some such and use it in both assignments to make it clear what's going on here?
___
*[`pkg/sql/bulkingest/split.go` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156877#-OdKuWqZBLYSZWDZAH4W:-OdKuWqZBLYSZWDZAH4X:b-y34m4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bb58a611be76c7dfc26a18379031ef673d00485/pkg/sql/bulkingest/split.go#L28)):*
> ```Go
> 	}
> 	for _, span := range spans {
> 		req := &kvpb.AdminScatterRequest{
> ```

Should this be db.AdminScatter()?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, and @mw5h)

___
*[`pkg/sql/bulkingest/split.go` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156877#-OdKuWqZBLYSZWDZAH4W:-OdOB0ZfA3IloCzjJ3be:bdpw90) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bb58a611be76c7dfc26a18379031ef673d00485/pkg/sql/bulkingest/split.go#L28)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Should this be db.AdminScatter()?
</blockquote></details>

I think this is correct. `db.AdminScatter` is designed for scattering a single key. Where as `kvpb.AdminScatterRequest` allows you to scatter a full span, which is what we want here.


<!-- Sent from Reviewable.io -->

mw5h reviewed (changes requested):
> @mw5h reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, and @spilchen)

___
*[`pkg/sql/bulkingest/split.go` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156877#-OdKuWqZBLYSZWDZAH4W:-OdOoXHsBERSbXTDmQwU:b22zn7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bb58a611be76c7dfc26a18379031ef673d00485/pkg/sql/bulkingest/split.go#L28)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I think this is correct. `db.AdminScatter` is designed for scattering a single key. Where as `kvpb.AdminScatterRequest` allows you to scatter a full span, which is what we want here.
</blockquote></details>

Sure, but don't we want to delegate this instead of consing up a raw kv request? It'd be trivial to introduce db.AdminScatterSpan() and then refer to it from db.AdminScatter().


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, and @mw5h)

___
*[`pkg/sql/bulkingest/split.go` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156877#-OdKuWqZBLYSZWDZAH4W:-OdP6HFQ2Zl6rGC-GE-f:b-7orfti) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bb58a611be76c7dfc26a18379031ef673d00485/pkg/sql/bulkingest/split.go#L28)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Sure, but don't we want to delegate this instead of consing up a raw kv request? It'd be trivial to introduce db.AdminScatterSpan() and then refer to it from db.AdminScatter().
</blockquote></details>

Ah ok. I see what you mean. I agree that's a worthwhile helper to add.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, @msbutler, and @mw5h)



<!-- Sent from Reviewable.io -->

msbutler commented on pkg/sql/bulkingest/split_picker.go:
```diff
@@ -0,0 +1,143 @@
+
+	for i := 1; i < len(ssts); i++ {
+		splitPoint := roachpb.Key(ssts[i].StartKey)
+		if safeSplitPoint, err := keys.EnsureSafeSplitKey(splitPoint); err == nil && len(safeSplitPoint) != 0 {
```
Also, `EnsureSafeSplitKey` could return a key that breaks the output span invariants:

> // The output spans cover the entire input span, are non-overlapping, and
// are contiguous. Each output span is assigned exactly one SST.

msbutler commented on pkg/sql/bulkingest/split.go:
```diff
@@ -0,0 +1,33 @@
+		}
+	}
+	for _, span := range spans {
+		if _, err := db.AdminScatterSpan(ctx, span); err != nil {
```
how will this function get called in actual source code? Will there be any parallelization? I ask because the `span` you scatter should be properly split at its start and end keys _before_ the scatter. Restore was plagued by a race related to this fact: https://github.com/cockroachdb/cockroach/issues/146083

spilchen replied:
As far as a I can tell (or at least how it was done in the prototype), is that it will be called sequentially at the cluster level. In the prototype, it was called by the IMPORT processor. So, no parallelism yet.  
<!-- Reviewable comment -OdxQnC22Fq0bvIs9SyM:b3p13xc -->
<!-- Sent from Reviewable.io -->

spilchen commented on pkg/sql/bulkingest/split_picker.go:
```diff
@@ -0,0 +1,143 @@
+
+	spanStart := span.Key
+
+	for i := 1; i < len(ssts); i++ {
```
I think they are expected to be about 128MB each. There is a knob to control this (see bulkio.sst_writer.batch_size)
<!-- Reviewable comment -OdxPsULCKzqftg_BXzv:bvlx6qn -->
<!-- Sent from Reviewable.io -->

spilchen replied:
I will merge this code as-is so that I don't block other aspects of the project. I have opened #158018 as a follow-on to use the simplified model.
<!-- Reviewable comment -OeNpGH98WhOZ_FcA1sb:b-u5d67v -->
<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, @msbutler, and @mw5h)



<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, @msbutler, and @mw5h)



<!-- Sent from Reviewable.io -->

mw5h reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @mw5h reviewed 2 of 2 files at r3, 1 of 2 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877#-:-OeJ09nY11HI5ucwSNHP:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, and @msbutler)



<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, and @msbutler)



<!-- Sent from Reviewable.io -->

jeffswenson reviewed (approved):
LGTM

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156877)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, and @msbutler)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTRs!

bors r+

spilchen commented:
bors retry

spilchen committed:
sql/bulkingest: add range split and scatter utilities

Adds split and scatter utilities for distributed merge from the
prototype branch. These utilities prepare the KV keyspace for efficient
bulk ingest by pre-splitting ranges at SST boundaries and distributing
them across cluster nodes.

Closes #156574
Epic: CRDB-48845
Release note: none

Co-authored by: @jeffswenson
